### ISSUE TITLE
Testing Rive with WidgetTester

### ISSUE THREAD

**Issue by andersunloc** [rive-app/rive-flutter#162]
Hi, I'm trying to render the final frame of a Rive animation with WidgetTester.

### Description
I'm trying to move the animation to the end by;
-setting time on the SimpleAnimation.instance.time
-setting speed to 0 at SimpleAnimation.instance.animation.speed
-advancing the animation with SimpleAnimation.instance.advance
-disabling / enabling animation with SimpleAnimation.isActive

It seems like whenever the animation start as isAction = false, it renders the first frame, and I dont get expected results from setting time to endTime, and setting the animation as not active.

How is this supposed to be done? That is; render a paused animation at last frame in WidgetTester?

**Comment by tp** [comment]
@andersunloc Did you manage to get testing working?

I also see an occasional (~30%) `pumpAndSettle timed out` error when running a golden test for a widget containing a rive animation (it's a simple UI control and should eventually be in a settled state, but alas Flutter thinks it does not…).

**Comment by moribellamy** [comment]
I'd also like to understand the best practice of using WidgetTester with rive. My case is that I have animations in a loop, but in test I don't care about them and would like `pumpAndSettle` to return after my microtask queue is empty.

I personally think there is enough confusion here to warrant an extra bit of documentation somewhere, re `Testing Rive with WidgetTester`. I'd love to know a "best practice" rather than use my questionable home-grown workarounds.

**Comment by moribellamy** [comment]
thanks @tp. i actually did something similar. if `isTest`, i just return `SizedBox.shrink()` instead of the rive object under test 😂 . none of my business logic depends on the successful completion of any animation, so i suspect this doesn't work for everyone. and it certainly seems smelly enough that i want to ask the community about it.

**Comment by HayesGordon** [comment]
Hi @andersunloc the animation will only be drawn on the next Flutter frame. So you can use `WidgetsBinding.instance.addPostFrameCallback` to pause your Rive animation. Here is an example (with two separate options) to get the last animation frame.


```dart
import 'package:flutter/material.dart';
import 'package:rive/rive.dart';

class RiveAnimationTest extends StatefulWidget {
  const RiveAnimationTest({super.key});

  @override
  State<RiveAnimationTest> createState() => _RiveAnimationTestState();
}

class _RiveAnimationTestState extends State<RiveAnimationTest> {
  SimpleAnimation controller = SimpleAnimation('Animation');
  late Artboard artboard;

  bool isTest = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: RiveAnimation.asset(
        "assets/rive/flux_capacitor.riv",
        controllers: [controller],
        onInit: (a) {
          artboard = a;
          if (isTest) {
            final linearAnimation = artboard.animationByName('Animation');
            // OPTION 1
            controller.apply(
                artboard as RuntimeArtboard,
                getLastFrameDuration(linearAnimation!.animation.durationSeconds,
                    linearAnimation.animation));

            // OPTION 2 - subtract a small number from durationSeconds
            // controller.apply(artboard as RuntimeArtboard,
            //     linearAnimation!.animation.durationSeconds - (1 / 100000));

            // Delay pausing the controller to the next Flutter frame to apply
            // animation.
            WidgetsBinding.instance.addPostFrameCallback((_) {
              controller.isActive = false;
            });
          }
        },
      ),
    );
  }
}

// Calculate the number of frames and subtract one frame.
double getLastFrameDuration(double value, LinearAnimation selectedAnimation) {
  var totalFrames = (value * selectedAnimation.fps)
      .clamp(0, selectedAnimation.duration)
      .roundToDouble();

  return (totalFrames - 1) / selectedAnimation.fps;
}
```

**Comment by HayesGordon** [comment]
@moribellamy we definitely can do with some Rive samples/documentation on how to do testing. I'm interested to hear thoughts on this and what type of examples the community would like to see?

**Comment by HayesGordon** [comment]
Some example code for a golden test:

```dart
import 'package:flutter/material.dart';
import 'package:flutter_rive/rive_animation_test.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Rive - Golden tests', () {
    testWidgets('Last frame test', (WidgetTester tester) async {
      await tester.pumpWidget(const MaterialApp(
        home: RiveAnimationTest(),
      ));
      await tester.pumpAndSettle();
      await expectLater(find.byType(RiveAnimationTest),
          matchesGoldenFile('rive-last-frame.png'));
    });
  });
}
```

**Comment by HayesGordon** [comment]
@tp that is interesting that you're running into the pumpAndSettle problem. Is it a complex Rive file, or does it have a lot of image assets in it?

**Comment by moribellamy** [comment]
> I'm interested to hear thoughts on this and what type of examples the community would like to see?

I think even the examples that were posted ad-hoc on this thread would be helpful. For example, now that I've seen an example from a Rive employee that uses an `isTest` boolean, i can rest somewhat assured that using that pattern is occasionally a best practice. also, your trick with `WidgetsBinding.instance.addPostFrameCallback` is good to know, and i'm not sure where else i could have learned that?

**Comment by tp** [comment]
> @tp that is interesting that you're running into the pumpAndSettle problem. Is it a complex Rive file, or does it have a lot of image assets in it?

No, not at all. It's a simple checkbox with a state machine (an initial animation depending on the first state, and then it would only animation when it's toggled again).

[checkbox.riv.zip](https://github.com/rive-app/rive-flutter/files/10801845/checkbox.riv.zip)

**Comment by tp** [comment]
@moribellamy I tried to apply your suggestion to my Rive-based widget, but it's still running into `pumpAndSettle` timeouts.

I suppose it works a bit different because of the state machine.

Here's the code for the above animation: https://gist.github.com/tp/dfee3d8f510bafdacdb5e71513d46459

The workaround  does not have the desired effect. Also even with `controller.isActive = false` I can see that the animation still updates after the toggle. So I suppose "disabling" the state machine controller does not have the desired effect.
I tried doing it on the animation level, but that did not have any effect either (I suppose because the state machine triggers them anew or something).

**Comment by moribellamy** [comment]
@tp looks like you may have used @HayesGordon 's suggestion re: the post-frame callback? I don't think I have the knowledge to help you.

The only success I've had is to just not render Rives in test at all, but obviously that cannot work for everyone's app.

**Comment by tp** [comment]
> @tp looks like you may have used @HayesGordon 's suggestion re: the post-frame callback? I don't think I have the knowledge to help you.

I tried that, but it did not work for the file I posted. I suppose it's related to the state machine, and that I have not found a way to run the _animations started by the state machine_ to completion/rest.
 
> The only success I've had is to just not render Rives in test at all, but obviously that cannot work for everyone's app.

Well, this would be a blocker for us to use Rive in more places, as we really need the fidelity of golden tests to ensure that everything fits together. For example we also use a bunch of color overrides on certain assets, and thus it's really important to have them rendered automatically to catch any unintended changes early.

**Comment by davidsdearaujo** [comment]
You can mock the assets data:
```dart
ServicesBinding.instance.defaultBinaryMessenger.setMockMessageHandler('flutter/assets', null);
```

**Comment by HayesGordon** [comment]
Hi @tp, there seemed to have been an issue with state machine and tests that was related to the way that Rive was advancing it's animations each frame.

This should be resolved now in v0.11.14

If you're encountering issues with `pumpAndSettle` then it means your animation is either in a loop. This could be intentional (if it's meant to loop) or it may be an issue in the animation that needs to be resolved. If your Rive animation never settles, when you expect it to, please create a separate issue and provide us the .rev file so that we can test.

For anyone that want some examples on testing, take a look at the [golden test folder ](https://github.com/rive-app/rive-flutter/tree/master/test/goldens) in this repository.

Going to close this issue for now.
